### DISCUSSION TITLE
Improve LOAD_ATTR specializations

### DISCUSSION THREAD

**Discussion by sweeneyde** [faster-cpython/ideas#214] [Ideas]
I was trying to figure out some of the situations in which LOAD_ATTR specializations can fail.

Here is a list of specializations stats by benchmark, along with a closer look at of some of the lowest-hit-percentage benchmarks: https://gist.github.com/sweeneyde/a0d3abfc99308b93c1cee084684f90e0

It seems there could be a benefit in specializing something to do with the type_getattro (i.e., `type.__getattr(ibute)__` in python), as many lookups of class attributes (whether from the class directly or indirectly through an instance) are not currently specialized.

**Comment by markshannon** [comment]
When you say something to do with `type.__getattribute__`, do you mean for attributes of the class access via an instance or on the class directly, or both?
i.e.
```
class C:
     a = NonDescriptor()
c = C()

c.a # This?
C.a # Or this? (or both?)
```

Two questions regarding the other failures:
1. Are we missing any important failure modes in our `SPEC_FAIL_...` list?
2. Do any of the failure modes in the `SPEC_FAIL_...` list not readily specialized, that could be if we split them up differently?

**Reply by sweeneyde** [reply]
> attributes of the class access via an instance or on the class directly, or both?

Both:

* Sympy benchmarks have a decent amounts of:
    * `Class.class_attr`
    * `Class.__name__` (data descriptor)
    * `Class.some_static_method`
    * `instance.class_attribute_of_some_superclass`

* Crypto_pyaes and scimark_monte_carlo:
    * failures are mostly `instance.class_attribute`

Regarding questions 1 and 2, right now SPEC_FAIL_OVERRIDDEN is too broad: it currently only says that `type->tp_getattro != PyObject_GenericGetAttr`. It would be good to distinguish (and probably specialize) the case `type->tp_getattro == PyType_Type.getattro`.

There's also the case for `instance.class_attr` where `_PyDictKeys_StringLookup(keys, name);` returns `DKIX_EMPTY`, so `index != (uint16_t)index`. Currently, this leads to SPEC_FAIL_OUT_OF_RANGE, which could be a bit misleading, and perhaps there could be further specializations for when the key is absent from the instance dict but present in the class dict.

**Reply by Fidget-Spinner** [reply]
For anyone interested, here's my (mostly dead) attempt at ``Class.class_attr`` specialization https://github.com/python/cpython/pull/29146. I tried two different approaches (one using `tp_version_tag` of `Class`, another using `tp_mro`'s ID). https://github.com/python/cpython/pull/29146#issuecomment-949409223

Some shortcomings:
1. Doesn't seem to specialize well on the standard test suite, not sure about pyperformance.
2. Uses more memory (I'm sure given enough time, someone smarter than me will have a more elegant solution which fixes this :)
